Refactor/Preparation for v1.0.0 release

carpawell opened the pull request:
Related to #170.

Contains changes in:
- checksum;
- container;
- object;
- storagegroup.

cthulhu-rider commented on container/announcement.go:
```diff
@@ -2,76 +2,97 @@ package container
+	*m = (container.UsedSpaceAnnouncement)(a)
 }
 
 // Epoch of the announcement.
```
I propose to not consider method name as a word:
```suggestion
// Epoch returns NeoFS epoch when UsedSpaceAnnouncement was made.
```
See other methods too.

cthulhu-rider commented on container/attribute.go:
```diff
@@ -4,75 +4,99 @@ import (
-)
-
-// NewAttribute creates and initializes blank Attribute.
+// Attribute represents v2-compatible container attribute.
```
```suggestion
// Attribute represents container attribute.
```

carpawell replied:
why? thought all public types should have such info

cthulhu-rider commented on container/attribute.go:
```diff
@@ -4,75 +4,99 @@ import (
-	(*container.Attribute)(a).SetKey(v)
-}
+// Attributes groups container attributes.
+type Attributes []Attribute
```
Do we really need this type?

carpawell replied:
think that it is good to have such since `object` has the same

also, it could be useful to convert to/from `v2` types and not do `for` loops in dapps

cthulhu-rider replied:
I agree that it is useful, but I propose to encapsulate all slice properties in methods:
```
// _ = Attributes([]Attribute{}) // not recommended

func (x Attributes) Len() int { ... }

func (x Attributes) Iterate(func(Attribute) bool) { ... }

func (x *Attributes) Append(...Attribute) { ... }

// etc.
```

What do you think? Otherwise IMO it's easier to work with `[]Attribute` directly.

carpawell replied:
sure, added

cthulhu-rider commented on container/attribute.go:
```diff
@@ -4,75 +4,99 @@ import (
-	for i := range v {
-		attrs[i] = *NewAttributeFromV2(&v[i])
+// ReadFromV2 reads Attributes from the []container.Attribute.
+// Attributes slice must not be nil.
```
Did you mean `Attributes`? Otherwise if u meant parameter then statement differs with the implementation.

carpawell replied:
hmmm, it was copied from other methods, but yes it is the senseless comment for that method, removed

cthulhu-rider commented on container/container.go:
```diff
@@ -14,6 +14,12 @@ import (
 	"github.com/nspcc-dev/neofs-sdk-go/version"
 )
 
+// Container is a structure that defines object placement behaviour.
```
```suggestion
// Container represents in-memory descriptor of the NeoFS container.
```

cthulhu-rider commented on container/container.go:
```diff
@@ -14,6 +14,12 @@ import (
+// Container is mutually compatible with github.com/nspcc-dev/neofs-api-go/v2/container.Container
+// message. See ReadFromV2 / WriteToV2 methods.
+//
+// Instances must be created via New function.
```
Too strong requirement, instances can be created using `var` declaration. For some scenarios we can provide analogue of `object.InitCreation`.

carpawell replied:
so you mean deleting `New` func or mentioning that var declaration is also ok?

cthulhu-rider replied:
Yes, we should refuse `New` functions and provide types with instances which can be safely instantiated using `var`. If there are some business logic like `each container must have owner` we implement it as a function like `InitCreation`.

carpawell replied:
ok, do not really see procs of the `InitCreation` function like in the `obj` service so fixed the comment and *just rename* `New` to `InitCreation` (do not get the idea of creation obj/cnr with any fields(1 TB payload, any address, any signature, etc.) pass it to `InitCreation` func and... expecting what?)

cthulhu-rider replied:
https://github.com/nspcc-dev/neofs-sdk-go/blob/f18b4a2a756bc49f26bde1bf202fc25c3daf2413/object/object.go#L19

https://github.com/nspcc-dev/neofs-sdk-go/blob/f18b4a2a756bc49f26bde1bf202fc25c3daf2413/object/object.go#L35

carpawell replied:
so we leave it as it is done no, yes?

my idea was to make an exception for such func (do not change passed object but return *really* blank object from the func since we can control that). but I do not insist

cthulhu-rider commented on container/container.go:
```diff
@@ -22,6 +28,25 @@ type Container struct {
 
+// ReadFromV2 reads Container from the container.Container message.
+//
+// Does not perform if message meets NeoFS API V2 specification.
```
Didn't catch the meaning, could you pls explain it?

carpawell replied:
it was copied from somewhere, changed to "Does not verify if the message meets NeoFS API V2 specification."

cthulhu-rider commented on container/container.go:
```diff
@@ -58,136 +83,169 @@ func New(opts ...Option) *Container {
-func (c *Container) ToV2() *container.Container {
+// CalculateID calculates container identifier
+// based on its structure.
+func CalculateID(c *Container) *cid.ID {
```
We must make it a method since versioning is encapsulated in `Container` type.

carpawell replied:
did not get your point, what does "versioning" mean? all methods called in the body of the `CalculateID` are exported

cthulhu-rider commented on container/container.go:
```diff
@@ -58,136 +83,169 @@ func New(opts ...Option) *Container {
 }
 
-func (c *Container) Version() *version.Version {
+// Version returns container's version.
```
This statement can be interpreted as a revision of a specific container.
```suggestion
// Version returns version of the NeoFS API protocol by which the container was created.
```

cthulhu-rider commented on container/container.go:
```diff
@@ -58,136 +83,169 @@ func New(opts ...Option) *Container {
+// Version must not be nil.
+//
+// See also Version.
 func (c *Container) SetVersion(v *version.Version) {
```
Accept value here and everywhere.

carpawell replied:
https://github.com/nspcc-dev/neofs-sdk-go/pull/186#discussion_r835918884

cthulhu-rider commented on container/container.go:
```diff
@@ -58,136 +83,169 @@ func New(opts ...Option) *Container {
 
-func (c *Container) BasicACL() uint32 {
-	return c.v2.GetBasicACL()
+// BasicACL returns container's ACL.
```
```suggestion
// BasicACL returns container's basic ACL.
```

cthulhu-rider commented on container/container.go:
```diff
@@ -58,136 +83,169 @@ func New(opts ...Option) *Container {
 
-func (c *Container) Attributes() Attributes {
-	return NewAttributesFromV2(c.v2.GetAttributes())
+// Attributes return container's attributes.
```
```suggestion
// Attributes returns container's attributes.
```

cthulhu-rider commented on container/container.go:
```diff
@@ -58,136 +83,169 @@ func New(opts ...Option) *Container {
 }
-
-// Marshal marshals Container into a protobuf binary form.
-func (c *Container) Marshal() ([]byte, error) {
```
Really unused?

carpawell replied:
if fact we discussed a little with @alexvanin that we gonna drop all marshaling methods and return them when doing SDK adapting since it is hard to track what is (un)used. but returned for container

as for my opinion, I'm still against saying that something *in a repo called SDK* is used and something is unused. for now, we are only able to say "it is unused in the node repo and the gateways", but this is not a library for them, somehow or other, we named it `SDK`

carpawell replied:
yea, i understand your point, but can not agree with "unused code" term in `SDK` repo

> support of backward compatibility

can not imagine problems with marshaling structs (I mean anyway we have (and will always have) such methods for some structs. so we have to support them and I do not see problems with supporting marshaling for all of our structs. if the question is about the amount of code that must be changed if something is changed in the marshaling, I think that the current refactor is more labor-intensive but anyway we have agreed with that)

cthulhu-rider commented on container/doc.go:
```diff
@@ -0,0 +1,55 @@
+/*
+Package container provides primitives to perform container operations in NeoFS.
+
+container type provides functionality to work with containers. For example
```
```suggestion
Container type provides functionality to work with containers. For example
```

cthulhu-rider commented on container/doc.go:
```diff
@@ -0,0 +1,55 @@
+
+Package also contains:
+
+• types for internals of the Object (Attribute, Attributes);
```
```suggestion
• types for internals of the Container (Attribute, Attributes);
```

cthulhu-rider commented on container/id/id.go:
```diff
@@ -10,39 +10,41 @@ import (
 )
 
 // ID represents v2-compatible container identifier.
```
```suggestion
// ID represents container identifier.
```

carpawell replied:
why? thought all public types should have such info

cthulhu-rider commented on container/id/id.go:
```diff
@@ -10,39 +10,41 @@ import (
-
 // Equal returns true if identifiers are identical.
-func (id *ID) Equal(id2 *ID) bool {
+func (id ID) Equal(id2 *ID) bool {
```
```suggestion
func (id ID) Equals(id2 ID) bool {
```

cthulhu-rider commented on container/id/id.go:
```diff
@@ -65,26 +69,11 @@ func (id *ID) Parse(s string) error {
+// Zero ID has returns empty string.
+//
+// See also Parse.
+func (id ID) String() string {
```
Lets replace it with `encoding.TestMarshaler` (same for `Parse`) and leave `fmt.Stringer` formatless.

carpawell replied:
https://github.com/nspcc-dev/neofs-sdk-go/pull/186#discussion_r835918936

cthulhu-rider commented on object/address/address.go:
```diff
@@ -10,6 +10,14 @@ import (
 )
 
 // Address represents v2-compatible object address.
```
```suggestion
// Address defines the address space for NeoFS objects.
```

carpawell replied:
why? thought all public types should have such info

cthulhu-rider commented on object/address/doc.go:
```diff
@@ -0,0 +1,9 @@
+/*
+Package address provides primitives to work with full (container and object
```
Pls don't mention `full` with object address since there are no such property.

cthulhu-rider commented on object/attribute.go:
```diff
@@ -5,69 +5,61 @@ import (
 )
 
 // Attribute represents v2-compatible object attribute.
```
```suggestion
// Attribute represents object attribute.
```

carpawell replied:
why? thought all public types should have such info

cthulhu-rider commented on object/error.go:
```diff
@@ -1,5 +1,7 @@
 package object
 
+// SplitInfoError is an error that appears when
```
```suggestion
// SplitInfoError is an error returned on split object encounter.
```

cthulhu-rider commented on object/error.go:
```diff
@@ -10,10 +12,13 @@ func (s *SplitInfoError) Error() string {
 
+// NewSplitInfoError creates new SplitInfoError
+// with provided SplitInfo.
 func NewSplitInfoError(v *SplitInfo) *SplitInfoError {
```
Lets provide setter and remove constructor.

carpawell replied:
delete even constructor for an error? i do not agree with that

most use cases are just returning `SplitInfoError` if we faced `SplitInfo`, so i do not see any preferences for using 3 lines of code (`var`, setter, `return`) instead of one (just returning with errConstructor): can not imaging that we could reuse one var via multiple setter calls

cthulhu-rider replied:
Then I suggest changing the constructor return to `error`.

cthulhu-rider commented on object/fmt.go:
```diff
@@ -86,14 +91,22 @@ func VerifyID(obj *Object) error {
 }
 
+// CalculateIDSignature signs object id with provided key.
 func CalculateIDSignature(key *ecdsa.PrivateKey, id *oid.ID) (*signature.Signature, error) {
```
Should be method same as container ID.

carpawell replied:
still do not get it

whose method should it be in *that* case?

cthulhu-rider commented on object/fmt.go:
```diff
@@ -105,10 +118,14 @@ func CalculateAndSetSignature(key *ecdsa.PrivateKey, obj *Object) error {
 }
 
+// VerifyIDSignature verifies object ID signature.
 func VerifyIDSignature(obj *Object) error {
```
Should be the method (lets return `bool` instead of `error`).

carpawell replied:
> Should be the method

still do not get it

> lets return bool instead of error

agree

carpawell replied:
fixed

cthulhu-rider commented on object/id/id.go:
```diff
@@ -11,25 +11,31 @@ import (
 )
 
 // ID represents v2-compatible object identifier.
```
```suggestion
// ID represents NeoFS object identifier.
```

carpawell replied:
why? thought all public types should have such info

cthulhu-rider commented on object/id/id.go:
```diff
@@ -38,18 +44,9 @@ func (id *ID) SetSHA256(v [sha256.Size]byte) {
-// Nil ID converts to nil.
-func (id *ID) ToV2() *refs.ObjectID {
-	return (*refs.ObjectID)(id)
+func (id ID) Equal(id2 *ID) bool {
```
```suggestion
func (id ID) Equals(id2 ID) bool {
```

cthulhu-rider commented on object/lock.go:
```diff
@@ -54,8 +57,11 @@ func (x *Lock) WriteMembers(ids []oid.ID) {
+
 		for i := range ids {
-			members[i] = *ids[i].ToV2() // need smth better
+			ids[i].WriteToV2(&idV2)
```
```suggestion
			ids[i].WriteToV2(&members[i])
```

cthulhu-rider commented on object/object.go:
```diff
@@ -13,12 +15,16 @@ import (
 )
 
 // Object represents in-memory structure of the NeoFS object.
```
// Object represents in-memory descriptor of the NeoFS object.

cthulhu-rider commented on object/object.go:
```diff
@@ -13,12 +15,16 @@ import (
+// Object is mutually compatible with github.com/nspcc-dev/neofs-api-go/v2/object.Object
+// message. See ReadFromV2 / WriteToV2 methods.
+//
+// Note that direct typecast is not safe and may result in loss of compatibility:
```
Lets place it at the bottom.

cthulhu-rider commented on object/object.go:
```diff
@@ -38,27 +44,19 @@ func InitCreation(dst *Object, rf RequiredFields) {
-}
-
-// New creates and initializes blank Object.
+// ReadFromV2 reads StorageGroup from the object.Object message.
```
```suggestion
// ReadFromV2 reads Object from the object.Object message.
```

cthulhu-rider commented on object/object.go:
```diff
@@ -38,27 +44,19 @@ func InitCreation(dst *Object, rf RequiredFields) {
-// into JSON format.
-func (o *Object) MarshalHeaderJSON() ([]byte, error) {
-	return (*object.Object)(o).GetHeader().MarshalJSON()
+// WriteToV2 writes StorageGroup to the object.Object message.
```
```suggestion
// WriteToV2 writes Object to the object.Object message.
```

cthulhu-rider commented on object/range.go:
```diff
@@ -5,47 +5,61 @@ import (
 )
 
 // Range represents v2-compatible object payload range.
```
```suggestion
// Range represents payload range of the NeoFS object.
```

carpawell replied:
why? thought all public types should have such info

cthulhu-rider replied:
About V2? We say about that below.

cthulhu-rider commented on object/search.go:
```diff
@@ -1,58 +1,79 @@
 	"github.com/nspcc-dev/neofs-sdk-go/version"
 )
 
 // SearchMatchType indicates match operation on specified header.
```
```suggestion
// SearchMatchType indicates match operation for selectable object properties.
```

cthulhu-rider reviewed (commented):
@carpawell pls open per-package PRs next time.

carpawell commented on checksum/checksum.go:
```diff
@@ -76,42 +93,22 @@ func (c *Checksum) SetTillichZemor(v [64]byte) {
-
+// Equal return boolean value that means
+// the equality of the passed Checksums.
 func Equal(cs1, cs2 *Checksum) bool {
```
as I understand, we decided not to change args and return values, only receivers

@alexvanin, @fyrchik, let's discuss if we gonna do it in another way

carpawell commented on checksum/checksum.go:
```diff
@@ -76,42 +93,22 @@ func (c *Checksum) SetTillichZemor(v [64]byte) {
-
-func (c *Checksum) String() string {
-	return hex.EncodeToString((*refs.Checksum)(c).GetSum())
+// String return Checksum string representation
```
> formatless

does it mean not to document `String` methods at all? what format such method should return?

cthulhu-rider replied:
I mean if type has fixed string format, it'd be better to implement `encoding.TextMarshaler/Unmarshaler` and make `String()` free and lightweight since it is used mostly in prints. About documenting I'd just write:
```
// String implements fmt.Stringer interface method.
```

carpawell replied:
i just do not understand what "formatless" means, do you mean that it should not be documented and it should be ok to change the format?

and also I do not get do you prefer leaving it as `HexString` method or implementing `encoding.TextMarshaler`

carpawell commented on checksum/checksum.go:
```diff
@@ -76,42 +93,22 @@ func (c *Checksum) SetTillichZemor(v [64]byte) {
 // Parse parses Checksum from its string representation.
+//
+// See also String.
 func (c *Checksum) Parse(s string) error {
```
do you mean to rename all `String` and `Parse` methods to `MarshalText` and `UnmarshalText` in the whole SDK?

@alexvanin, @fyrchik let's discuss it

cthulhu-rider replied:
Yes, we've started to do this some time ago, but code migration aborted the process.

carpawell commented on checksum/doc.go:
```diff
@@ -0,0 +1,18 @@
+/*
+Package checksum provides primitives to work with sums/hashes of
```
somewhere we name it "checksum" and somewhere it is "hash" so I decided to document both

carpawell commented on client/object_search.go:
```diff
@@ -129,8 +129,11 @@ func (x *ObjectListReader) Read(buf []oid.ID) (int, bool) {
 	for i := 0; i < read; i++ {
-		buf[i] = *oid.NewIDFromV2(&x.tail[i]) // need smth better
+		id.ReadFromV2(x.tail[i])
+		buf[i] = id // need smth better
```
my bad, fixed

carpawell commented:
> pls open per-package PRs next time.

@cthulhu-rider, yeah, I would try that next time, did not expect that PR to be that big, changes in one pkg lead to changes in another pkg

carpawell commented:
- returned deleted `Marshal/...` funcs (did not add a new ones);
- changed params to structs (not pointers);
- all `FromString` renamed to `Parse`;
- all `String` methods are now documented as "String implements fmt.Stringer interface method.";
- added `Empty()` methods to all ids and to some other optional structs.

cthulhu-rider commented:
Too hard to support, will be reopened later.

cthulhu-rider commented:
Too hard to support, will be reopened later.
